### ISSUE TITLE
Are two generator passes per training step actually necessary?

### ISSUE THREAD

**Issue by kklemon** [akanimax/msg-stylegan-tf#11]
I'm currently working on an own implementation of your great paper during which I came up with the question, if the second generator pass in each training step is actually necessary. What I mean is, that we generate fake samples once to optimize D and then once again to do so for G. But for the second case, the weights of G actually haven't changed so the drawn fakes won't differ either. It would be different, if we used different latent samples for each pass, but this does not seem to be the case.

From my current understanding, we could simply retain the gradients from the first pass and avoid one costly generator call, reducing the computational resources by 25% or something per training iteration. But I have to admit, that I'm more familiar with your PyTorch implementation of the paper, which I took as a reference, to clarify on this question. Nonetheless, I'm asking here, as it seems to be "official" repository.

**Comment by animeshkarnewar-tomtom** [comment]
Well, firstly, a good question :+1: and thanks for your kind words. This is indeed the official paper repo. Actually, we do sample latents differently for the D and G passes in this repository. It happens while computing the loss ([here](https://github.com/akanimax/msg-stylegan-tf/blob/master/training/loss.py#L23) for instance). You are right about the G weights not changing after the preceding D pass, but reason why we resample and regenerate is that for almost any distribution matching GAN loss, the expectation of the D scores is computed using monte-carlo approximation, so technically we should regenerate fake differently for the G pass. In practice though, from my experience, it kind of works even if you don't resample, although you do run a higher risk of mode collapse. Hope this helps.
Please feel free to ask if you have any further questions.

**Comment by kklemon** [comment]
That makes sense. Thank you for the reply! Indeed, from my experience resampling or not does not make a huge difference. I would even claim, that for models of this size, the aforementioned savings in computational requirements can maybe even outweigh the slightly reduced stability.
